Add neo-beta-1 support to v14.x.x

nemanja-boric-sociomantic opened the pull request:
(to release v14.1.x)

This adds back elementary support for neo to v14.x.x, then it rebases neo branch and finally does all
necessary changes to compile neo with swarm v5.x.x.

nemanja-boric-sociomantic committed:
Enhance DlsClient's constructors' documentation

nemanja-boric-sociomantic committed:
Fix reference to ClassFiller (change it to ConfigFiller)

It's renamed to ConfigFiller, so docs should be updated.

nemanja-boric-sociomantic committed:
Add DlsClient constructors accepting Config/NeoConfig instances

nemanja-boric-sociomantic committed:
Adapt usage examples to use config-accepting DlsClient ctor

nemanja-boric-sociomantic committed:
Use ConnectionHandler.RequestMap instead of deprecated RequestHandlers

nemanja-boric-sociomantic committed:
Update SuspendableRequest framework to use nextEvent

Although this is deprecated in the new swarm, it's worth keeping
it until we rewrite entire GetRange to use RequestEventDispatcher.

nemanja-boric-sociomantic committed:
Use SchedulingDlsClient.Config in neo SchedulingDlsClient constructor

This patch fixes an apparent oversight in providing the neo constructor
for `SchedulingDlsClient`: there is no need for a separate `max_events`
parameter, given that it is already provided in the class' own custom
`Config` class.

This should make it easier for users to transition to the neo protocol
as it maintains consistency with the other `Config`-using constructor
overloads of `SchedulingDlsClient`.

nemanja-boric-sociomantic committed:
Update submodules to latest versions

* submodules/ocean v3.3.1(a991c41)...v3.5.5(1047765) (134 commits)
  > Merge tag v3.4.6 into v3.5.x
  > Merge tag v3.4.5 into v3.5.x
  > Fix DaemonApp.onSignal crash on unkown signal
  > Add missing imports in deprecated modules
  > Merge tag v3.4.4 into v3.5.x
  (...)

* submodules/swarm v4.3.2(00fddf8)...v4.4.0(e4ab31c) (22 commits)
  > Time neo requests for which timing stats are configured
  > Add test for neo stats logging without request timing
  > Add facility to disable timing stats for specific requests in test nodes
  > Allow node implementations to turn off timing stats for some neo requests
  > Don't log timing stats for requests without timing info
  (...)

* submodules/turtle v8.0.0(a319c42)...v8.2.1(b32db61) (18 commits)
  > Add a note to remove module ctor in v9.0.0
  > Fix turtle import affecting unittests
  > Fix broken merge of --memcheck feature
  > Collect peak virtual memory usage of the app
  > Refactor TestedApplicationBase kill timer
  (...)

nemanja-boric-sociomantic committed:
Fix documentation for dlstest.DlsClient.connectionNotifier

Fixes #33

Commit:
Remove unused imports

Commit:
Use TaskEvent to suspend/resume tasks in DlsClient test

It makes easier to read the code, debug and actually safer
if `TaskEvent` is used to suspend/resume a task rather than
using directly the plain suspend/resume from the task.

nemanja-boric-sociomantic committed:
Don't resume the Task on node_error in Blocking task support

nemanja-boric-sociomantic committed:
Migrate GetRange to multi-fiber-framework

This migrates the GetRange request from the
SuspendableRequest framework to the new multi-fiber framework.

nemanja-boric-sociomantic committed:
Remove GetRange v0 protocol

nemanja-boric-sociomantic committed:
Rename GetRange.MessageType to MessageType_v1

As we're going to support multiple request versions at once,
we need distinct enums for the identifiers between node/client.

nemanja-boric-sociomantic committed:
Add back support for GetRange_v0

Since the node needs to support multiple versions of the request
during the transition period, we need to get back the GetRange_v0 to
the node's implementation.

nemanja-boric-sociomantic committed:
Send GetRange.Finished on the node error

In case there's an error during the request, node will send Finished
message to the client, indicating that the request has finished. This is
transitional and non-ideal fix, since the client can't know that there
was an error during the request.

nemanja-boric-sociomantic committed:
Flush in the node after sending the batch and Stopped message

nemanja-boric-sociomantic committed:
Flush in client after sending Continue/Stop

Commit:
Update list of supported versions

nemanja-boric-sociomantic committed:
Use double-buffering to fetch the records from the one during iteration

The previous approach where the client would ask for the batch from the
node was prone to the latency problems - if the time to prepare and send
batch is non-negligible the performance of the client application would
suffer. This patch improves the situation and decreases the latency in
the following way:

- As soon as client get the new batch from the node, it will request another one
- When the new batch arrives the client will store it in the back
  buffer, ready for processing as soon as it finish with the previous one
- After processing the previous batch and the new batch already arrived,
  the client will just swap the buffers, reading from the new one.

nemanja-boric-sociomantic committed:
Append to the input batch, don't replace it

Since the node will in some circumstances send the data after the Stop
signal is resumed.

nemanja-boric-sociomantic committed:
Use this.buffers.empty() to figure if we should wait for more records

Instead of using `input` field directly, we should use the method which
tells us if the buffers are empty.

nemanja-boric-sociomantic committed:
Don't send the empty batch before Finished message

There's no point in sending the empty batch if it will be followed by
the Finished message. The finished message alone is enough to end the
request.

Fixes #62

nemanja-boric-sociomantic committed:
Shutdown the client with the protocol error on an empty batch

Under no circumstances should the client receive an empty batch from
the node. If that happens, just end the request on that connection.

nemanja-boric-sociomantic committed:
Throw exception returned by shutdownWithProtocolError in GetRange

In one place, the exception returned by `shutdownWithProtocol` error
wasn't thrown.

nemanja-boric-sociomantic committed:
Raise stopped notification if request ends as the result of the stop()

The stopped notification was missing, however code that was preventing
finished notification was in place, resulting in the user lacking both
stopped and finished notifications in case GetRange.stop() was used.

nemanja-boric-sociomantic committed:
Fix deprecations in v2.071.x in neo branch

nemanja-boric-sociomantic committed:
Remove GetRange_v0 from the Node's side

It was already removed from the client's side.

nemanja-boric-sociomantic committed:
Add DlsClient constructor accepting neo but not legacy configuration

In case legacy configuration is not needed to be configured through
the Config file, the new constructor may be used that still initializes
legacy protocol, but accepts config file only for the Neo.

nemanja-boric-sociomantic committed:
Fix the documentation to reflect the new ConnectionNotifier signature

nemanja-boric-sociomantic committed:
GetRange notification types started, suspended and resumed are removed

Since the GetRange v1, the request is immediately ready to be suspended,
and it also suspends and resumes instantly, without waiting for the
nodes.  This makes `started`, `suspended` and `removed` notification
types useless and they are now removed. The `started` notification
should be replaced with the code that follows the creation of the
request, and the code for suspend/resume should be removed and it should
be assumed that the client will suspend/resume immediately.

nemanja-boric-sociomantic committed:
Fix comment from UsageExamples wrt started/suspended/resumed notification

nemanja-boric-sociomantic committed:
Revert "Remove outdated neo support"

This reverts commit a7b73a388386d4c17842852d54c1b77526a3e821.

nemanja-boric-sociomantic committed:
Bump the version of the requests

There's a breaking change in the request framework compared to the
neo-beta codes, so this changes the request versions.

nemanja-boric-sociomantic committed:
Remove suspendable request framework

nemanja-boric-sociomantic committed:
Make notifiers accept user arguments as const

nemanja-boric-sociomantic committed:
Update Neo support to swarm 5.0.1

nemanja-boric-sociomantic commented:
Blocked on https://github.com/sociomantic-tsunami/swarm/pull/323

Just last three commits are new, other is just rebased from `neo` branch + revert of neo removal.

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
     ***************************************************************************/
 
-    final public void handle ( RequestOnConn connection, Const!(void)[] msg_payload )
+    public void preSupportedCodeSent ( Const!(void)[] msg_payload )
```
It's recommended to use the new new request handler framework, which doesn't require these separate pre and post methods. See https://github.com/sociomantic-tsunami/swarm/blob/v5.x.x/src/swarm/neo/node/ConnectionHandler.d#L161.

nemanja-boric-sociomantic replied:
Err, but this hasn't been released yet - I've just tried to use it from swarm v5.0.x. Is it time for v5.1.0?

nemanja-boric-sociomantic replied:
Even with v5.1.0, I can't slip the swarm dependency upgrade in v14.x.x of dlsproto :-(.

gavin-norman-sociomantic replied:
Ahh sorry, I forgot it wasn't in v5.0.x.

gavin-norman-sociomantic reviewed (changes requested):
I think you also need to `verify` the rebased neo code.

nemanja-boric-sociomantic commented:
Ah, `verify` is a good point, thanks!

nemanja-boric-sociomantic committed:
Update swarm to v5.0.14

* submodules/swarm v5.0.0(508da81)...v5.0.14(a6a893c) (15 commits)
  > Merge tag v4.6.8 into v5.0.x
  > Verify the RequestSetup module
  > Merge tag v4.6.7 into v5.0.x
  > Merge tag v4.6.6 into v5.0.x
  > Merge tag v4.6.5 into v5.0.x
  (...)

nemanja-boric-sociomantic committed:
Convert asserts to verify in Neo DLS part

The newly introduced neo support should use verify to perform sanity
checks.

nemanja-boric-sociomantic committed:
Add release notes and a client readme

nemanja-boric-sociomantic commented:
Updated with the updated swarm (so this is green now), `assert` to `verify` conversion and added release notes and a client readme.

nemanja-boric-sociomantic commented:
And also a slight refactoring of the request handlers framework update commit (extracted `this.connection.event_dispatcher` to a `ed` field of a request handler).

nemanja-boric-sociomantic committed:
Fix visibility issues with DMD 2.071

gavin-norman-sociomantic commented on src/dlsproto/client/README.rst:
```diff
@@ -0,0 +1,60 @@
+DLS Client Overview
+===================
+
+This package contains the client to connect to a Distributed Log Storage
```
Officially it's Store, not Storage.

gavin-norman-sociomantic commented on src/dlsproto/client/README.rst:
```diff
@@ -0,0 +1,60 @@
+The DLS supports the following requests (links to the API modules):
+
+* `Put <request/Put.d>`_:
+  pushes a record to a channel.
```
Maybe "adds a record to a channel at the specified timestamp."?

gavin-norman-sociomantic commented on src/dlsproto/client/README.rst:
```diff
@@ -0,0 +1,60 @@
+* `Put <request/Put.d>`_:
+  pushes a record to a channel.
+* `GetRange <request/GetRange.d>`_:
+  receives a batch of records in the given range, optionally filtered by a
```
Something like "in the given timestamp range"?

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
-        finally
+    }
+
+    public void postSupportedCodeSent ()
```
Doc missing.

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
+        this.compressed_batch = this.resources.getVoidBuffer();
+        this.saved_exception = this.resources.getException();
+        this.lzo = this.resources.getLzo();
+
```
Double newline.

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/GetRange.d:
```diff
@@ -149,77 +126,83 @@ public abstract scope class GetRangeProtocol_v1
+
+    public void postSupportedCodeSent ()
+    {
+        this.value_buffer = this.resources.getVoidBuffer();
```
Small optimization: you could only acquire these resources if the request was initialised ok.

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/Put.d:
```diff
@@ -12,77 +12,70 @@
         }
     }
 
+    public void postSupportedCodeSent ()
```
Missing doc.

gavin-norman-sociomantic commented on src/dlsproto/node/neo/request/Put.d:
```diff
@@ -12,77 +12,70 @@
-        cstring channel = parser.getArray!(char)(msg_payload);
-        time_t timestamp = *parser.getValue!(time_t)(msg_payload);
-        auto value = parser.getArray!(char)(msg_payload);
+        cstring channel = this.ed.message_parser.getArray!(char)(msg_payload);
```
Btw David recommended using `parseBody`. See https://github.com/sociomantic-tsunami/dhtproto/pull/110#discussion_r187033167.

nemanja-boric-sociomantic replied:
Nice tip, I forgot that this was there.

nemanja-boric-sociomantic committed:
Use parseBody inside Put request instead of parsing individual items

If the message payload always carries the same set of fields, it's
better to use parseBody instead of individual `get*` methods.

nemanja-boric-sociomantic commented:
Updated with all comments addressed.

nemanja-boric-sociomantic commented:
Red because of sourceforge issues.

gavin-norman-sociomantic reviewed (approved):
LGTM. Feel free to auto-merge when CI is back up.
